README
NAME
    POE::Component::Metabase::Relay::Server - A Metabase relay server
    component

VERSION
    version 0.34

SYNOPSIS
      use strict;
      use warnings;

      use POE qw[Component::Metabase::Relay::Server];

      my $test_httpd = POE::Component::Metabase::Relay::Server->spawn(
        port    => 8080,
        id_file => shift,
        dsn     => 'dbi:SQLite:dbname=dbfile',
        uri     => 'https://metabase.example.foo/',
        debug   => 1,
      );

      $poe_kernel->run();
      exit 0;

DESCRIPTION
    POE::Component::Metabase::Relay::Server is a relay server for Metabase.
    It provides a listener that accepts connections from
    Test::Reporter::Transport::Socket based CPAN Testers and relays the
    Storable serialised data to Metabase using
    POE::Component::Metabase::Client::Submit.

    POE::Component::Client::HTTP is used to submit reports usually, but if
    version 0.06 of POE::Component::Curl::Multi is found to be installed,
    this will be used in preference. You can disable this usage using the
    "no_curl" option to "spawn".

CONSTRUCTOR
    "spawn"
        Spawns a new component session and creates a SQLite database if it
        doesn't already exist.

        Takes a number of mandatory parameters:

          'id_file', the file path of a Metabase ID file;
          'dsn', a DBI DSN to use to store the submission queue;
          'uri', the uri of metabase server to submit to;

        and a number of optional parameters:

          'address', the address to bind the listener to, defaults to INADDR_ANY;
          'port', the port to listen on, defaults to 0, which picks a random port;
          'username', a DSN username if required;
          'password', a DSN password if required;
          'db_opts', a hashref of DBD options that is passed to POE::Component::EasyDBI;
          'debug', enable debugging information;
          'multiple', set to true to enable the Queue to use multiple PoCo-Client-HTTPs, default 0;
          'no_relay', set to true to disable report submissions to the Metabase, default 0;
          'no_curl',  set to true to disable automatic usage of POE::Component::Curl::Multi, default 0;
          'submissions', an int to control the number of parallel http clients ( used only if multiple == 1 ), default 10;
          'session', a POE::Session alias or session ID to send events to;
          'recv_event', an event to be triggered when reports are received by the relay;

        "address" may be either an simple scalar value or an arrayref of
        addresses to bind to.

        If "recv_event" is specified an event will be sent for every report
        received by the relay server. Unless "session" is specified this
        event will be sent to the parent session of the component.

OUTPUT EVENTS
    If "recv_event" is specified to "spawn", an event will be sent with the
    following:

    "ARG0" will be a "HASHREF" with the following keys:

     osversion
     distfile
     archname
     textreport
     osname
     perl_version
     grade

    "ARG1" will be the IP address of the client that sent the report.

    If "queue_event" is specified to "spawn", an event will be sent for
    particular changes in queue status

AUTHOR
    Chris Williams <chris@bingosnet.co.uk>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Chris Williams.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
==================================================
Changes from 2009-03-13 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 0.34 at 2014-03-12 20:18:33 +0000
-----------------------------------------

  Change: 0826f3ff6bada096d840c39d3ba4d7b04fe7bdd4
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2014-03-12 20:18:33 +0000

    Fix no_curl option and require 0.08 of PoCo-Curl-Multi 

-----------------------------------------
version 0.32 at 2014-03-11 14:04:35 +0000
-----------------------------------------

  Change: 3aa21b8f60218ae5e0417674430263bee8083bfd
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2014-03-11 14:04:35 +0000

    Enable automatic usage of POE::Component::Curl::Multi if it is found.

    Add an option to disable this (no_curl) if required 

-----------------------------------------
version 0.30 at 2014-02-13 14:22:54 +0000
-----------------------------------------

  Change: 366fa9dc333999a85f6d03ec65c7e2c44857ccf6
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2014-02-13 14:22:54 +0000

    Resolve (huh) problems when multiple is set 

-----------------------------------------
version 0.28 at 2014-02-12 11:36:40 +0000
-----------------------------------------

  Change: 3e27cb4bd5f5b1de109ea8183a2e65bfbdcbdeef
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2014-02-12 11:36:40 +0000

    Remove requirement for MooseX::AttributeHelpers 

-----------------------------------------
version 0.26 at 2014-01-21 19:10:05 +0000
-----------------------------------------

  Change: 9a9bbd9864a57aa0da1146fc486f4c82b2016753
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2014-01-21 19:10:05 +0000

    Resolve issue with constraints and _set_resolver

    Tidy up stringification of id_file warnings 

-----------------------------------------
version 0.24 at 2011-11-17 22:00:44 +0000
-----------------------------------------

  Change: b189f0e317e83ecad803bb5fe8013a7ddb89243d
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2011-11-17 22:00:44 +0000

    Update required version of MooseX::Types::URI 

-----------------------------------------
version 0.22 at 2011-04-26 23:01:37 +0000
-----------------------------------------

  Change: bce1de1360ee41919392abd33b0cbc91d31e679d
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2011-04-27 00:01:37 +0000

    Bump version 

  Change: 0cc49e1cf7cd09120a9d0e3f095d849d7400272c
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2011-04-27 00:00:38 +0000

    Ensure that queue processing starts immediately on startup if there
    are backlogged requests 

-----------------------------------------
version 0.20 at 2011-03-08 13:27:15 +0000
-----------------------------------------

  Change: 691ea2436e7222cdd178eb2637c2c4592ff4d032
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2011-03-08 13:27:15 +0000

    Bump version to 0.20 

  Change: b5ccebd9469357f0d9c409faa4d4585664daa92d
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2011-03-08 13:25:46 +0000

    Update required versions of prereqs and add new prereqs. 

-----------------------------------------
version 0.18 at 2010-08-13 11:06:39 +0000
-----------------------------------------

  Change: 75d1ff9d208127acac6f31d0b3a05fa233898636
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2010-08-13 12:06:39 +0000

    Added extra debug info when we receive data we can't parse. Enqueued
    event generation to the queue. 

-----------------------------------------
version 0.16 at 2010-07-08 13:33:14 +0000
-----------------------------------------

  Change: 8879c31d6a5c5f0516c03c8de8407ed0f5681cdd
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2010-07-08 14:33:14 +0000

    fixed slight issues with the 'session' detection 

-----------------------------------------
version 0.14 at 2010-07-08 10:39:49 +0000
-----------------------------------------

  Change: cebaa2cee3a8a20a4c26c42861f237e8315f3691
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2010-07-08 11:39:49 +0000

    Fix the POD 

  Change: 647cfcfd41f21cf710d2687d0c65f8d5d9eb727c
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2010-07-08 11:24:57 +0000

    Added 'session' and 'recv_event' parameters 

  Change: d2f6d7bb4b35420058a715f3f2e9f7cad22a70d1
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2010-07-07 22:01:22 +0000

    Added test and an example metabase profile file, found and fixed
    errors. 

  Change: 44c32077bfe420acc5498e59559905195b06e3c6
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2010-07-07 12:50:33 +0000

    Switched distribution to Dist::Zilla 

-----------------------------------------
version 0.12 at 2010-04-30 14:40:37 +0000
-----------------------------------------

  Change: 000ec9a059208e41a6dc5ab08856397ac571a5c5
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2010-04-30 15:40:37 +0000

    Bump version to 0.12 

  Change: 8291bea87907e163157d6a6a644bae7ecd17ca09
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2010-04-30 15:39:14 +0000

    Sync up 

  Change: 3e16f27b007eaa8c7d84e135b6204a43ec24977b
  Author: Florian Ragwitz <rafl@debian.org>
  Date : 2010-04-30 16:15:14 +0000

    Stop importing all of Socket::. 

  Change: 5019807fa208b95c8d2d1a3d6536a79a11f0ea17
  Author: Florian Ragwitz <rafl@debian.org>
  Date : 2010-04-30 16:14:54 +0000

    Disable pod tests for users. 

-----------------------------------------
version 0.10 at 2010-04-28 23:41:20 +0000
-----------------------------------------

  Change: b9977e94b3ba4d045f55429014b5d08d82f6f8bc
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2010-04-29 00:41:20 +0000

    Bump version 


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.013.
use strict;
use warnings;

use 5.006000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A Metabase relay server component",
  "AUTHOR" => "Chris Williams <chris\@bingosnet.co.uk>",
  "BUILD_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.42",
    "Test::More" => "0.47"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.42"
  },
  "DISTNAME" => "POE-Component-Metabase-Relay-Server",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "POE::Component::Metabase::Relay::Server",
  "PREREQ_PM" => {
    "CPAN::Testers::Report" => "1.999",
    "DBD::SQLite" => "1.29",
    "DBI" => "1.609",
    "Data::UUID" => "1.203",
    "JSON" => "2.17",
    "JSON::XS" => "2.29",
    "Metabase::Fact" => "0.011",
    "Metabase::User::Profile" => "0.003",
    "Metabase::User::Secret" => "0.003",
    "Module::Load::Conditional" => 0,
    "Moose" => "1.01",
    "Moose::Util::TypeConstraints" => "1.01",
    "MooseX::POE" => "0.205",
    "MooseX::Types::Path::Class" => "0.05",
    "MooseX::Types::URI" => "0.03",
    "POE" => "1.299",
    "POE::Component::Client::HTTP" => "0.948",
    "POE::Component::Client::Keepalive" => "0.271",
    "POE::Component::EasyDBI" => "1.23",
    "POE::Component::Metabase::Client::Submit" => "0.12",
    "POE::Component::Resolver" => "0.921",
    "Params::Util" => "1.01",
    "Storable" => 0,
    "Test::POE::Server::TCP" => "1.12",
    "Time::HiRes" => 0,
    "URI" => "1.52",
    "base" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.47"
  },
  "VERSION" => "0.34",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::Testers::Report" => "1.999",
  "DBD::SQLite" => "1.29",
  "DBI" => "1.609",
  "Data::UUID" => "1.203",
  "ExtUtils::MakeMaker" => "6.42",
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "JSON" => "2.17",
  "JSON::XS" => "2.29",
  "Metabase::Fact" => "0.011",
  "Metabase::User::Profile" => "0.003",
  "Metabase::User::Secret" => "0.003",
  "Module::Load::Conditional" => 0,
  "Moose" => "1.01",
  "Moose::Util::TypeConstraints" => "1.01",
  "MooseX::POE" => "0.205",
  "MooseX::Types::Path::Class" => "0.05",
  "MooseX::Types::URI" => "0.03",
  "POE" => "1.299",
  "POE::Component::Client::HTTP" => "0.948",
  "POE::Component::Client::Keepalive" => "0.271",
  "POE::Component::EasyDBI" => "1.23",
  "POE::Component::Metabase::Client::Submit" => "0.12",
  "POE::Component::Resolver" => "0.921",
  "Params::Util" => "1.01",
  "Storable" => 0,
  "Test::More" => "0.47",
  "Test::POE::Server::TCP" => "1.12",
  "Time::HiRes" => 0,
  "URI" => "1.52",
  "base" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/POE/Component/Metabase/Relay/Server.pm
package POE::Component::Metabase::Relay::Server;
$POE::Component::Metabase::Relay::Server::VERSION = '0.34';
# ABSTRACT: A Metabase relay server component

use strict;
use warnings;
use CPAN::Testers::Report;
use POE qw[Filter::Stream];
use POE::Component::Metabase::Relay::Server::Queue;
use Test::POE::Server::TCP;
use Carp                      ();
use Storable                  ();
use Socket                    ();
use JSON                      ();
use Metabase::User::Profile   ();
use Metabase::User::Secret    ();

my @fields = qw(
  osversion
  distfile
  archname
  textreport
  osname
  perl_version
  grade
);

use MooseX::POE;
use MooseX::Types::Path::Class qw[File];
use MooseX::Types::URI qw[Uri];

{
  use Moose::Util::TypeConstraints;
  my $tc = subtype as 'ArrayRef[Str]';
  coerce $tc, from 'Str', via { [$_] };

  has 'address' => (
    is => 'ro',
    isa => $tc,
    default => 0,
    coerce => 1,
  );

  my $ps = subtype as 'Str', where { $poe_kernel->alias_resolve( $_ ) };
  coerce $ps, from 'Str', via { $poe_kernel->alias_resolve( $_ )->ID };

  has 'session' => (
    is => 'ro',
    isa => $ps,
    coerce => 1,
    writer => '_set_session',
  );

  no Moose::Util::TypeConstraints;
}

has 'port' => (
  is => 'ro',
  default => sub { 0 },
  writer => '_set_port',
);

has 'id_file' => (
  is       => 'ro',
  required => 1,
  isa      => File,
  coerce   => 1,
);

has 'dsn' => (
  is => 'ro',
  isa => 'Str',
  required => 1,
);

has 'uri' => (
  is => 'ro',
  isa => Uri,
  coerce => 1,
  required => 1,
);

has 'username' => (
  is => 'ro',
  isa => 'Str',
  default => '',
);

has 'password' => (
  is => 'ro',
  isa => 'Str',
  default => '',
);

has 'db_opts' => (
  is => 'ro',
  isa => 'HashRef',
  default => sub {{}},
);

has 'debug' => (
  is => 'rw',
  isa => 'Bool',
  default => 0,
);

has 'multiple' => (
  is => 'ro',
  isa => 'Bool',
  default => 0,
);

has 'recv_event' => (
  is => 'ro',
  isa => 'Str',
);

has 'no_relay' => (
  is => 'rw',
  isa => 'Bool',
  default => 0,
  trigger => sub {
    my( $self, $new, $old ) = @_;
    return if ! $self->_has_queue;
    $self->queue->no_relay( $new );
  },
);

has 'no_curl' => (
  is => 'ro',
  isa => 'Bool',
  default => 0,
);

has 'submissions' => (
  is => 'rw',
  isa => 'Int',
  default => 10,
  trigger => sub {
    my( $self, $new, $old ) = @_;
    return if ! $self->_has_queue;
    $self->queue->submissions( $new );
  },
);

has '_profile' => (
  is => 'ro',
  isa => 'Metabase::User::Profile',
  init_arg => undef,
  writer => '_set_profile',
);

has '_secret' => (
  is => 'ro',
  isa => 'Metabase::User::Secret',
  init_arg => undef,
  writer => '_set_secret',
);

has '_relayd' => (
  accessor => 'relayd',
  isa => 'ArrayRef[Test::POE::Server::TCP]',
  lazy_build => 1,
  auto_deref => 1,
  init_arg => undef,
);

has '_queue' => (
  accessor => 'queue',
  isa => 'POE::Component::Metabase::Relay::Server::Queue',
  lazy_build => 1,
  init_arg => undef,
);

has '_requests' => (
  is => 'ro',
  isa => 'HashRef',
  default => sub {{}},
  init_arg => undef,
);

sub _build__relayd {
  my $self = shift;
  return [map {
    Test::POE::Server::TCP->spawn(
        address => $_,
        port => $self->port,
        prefix => 'relayd',
        filter => POE::Filter::Stream->new(),
    )
  } @{ $self->address }]
}

sub _build__queue {
  my $self = shift;
  POE::Component::Metabase::Relay::Server::Queue->spawn(
    dsn      => $self->dsn,
    username => $self->username,
    password => $self->password,
    db_opts  => $self->db_opts,
    uri      => $self->uri->as_string,
    profile  => $self->_profile,
    secret   => $self->_secret,
    debug    => $self->debug,
    multiple => $self->multiple,
    no_relay => $self->no_relay,
    no_curl  => $self->no_curl,
    submissions => $self->submissions,
  );
}

sub spawn {
  shift->new(@_);
}

sub START {
  my ($kernel,$self,$sender) = @_[KERNEL,OBJECT,SENDER];
  if ( $kernel == $sender and $self->recv_event and !$self->session ) {
    Carp::croak "Not called from another POE session and 'session' wasn't set\n";
  }
  if ( $self->recv_event ) {
    $self->_set_session( $sender->ID ) unless $self->session;
  }
  $self->_load_id_file;
  $self->relayd;
  $self->queue;
  return;
}


event 'shutdown' => sub {
  my ($kernel,$self) = @_[KERNEL,OBJECT];
  $_->shutdown for $self->relayd;
  $poe_kernel->post(
    $self->queue->get_session_id,
    'shutdown',
  );
  return;
};

event 'relayd_registered' => sub {
  my ($kernel,$self,$relayd) = @_[KERNEL,OBJECT,ARG0];
  my ($port, $addr) = Socket::sockaddr_in($relayd->getsockname);
  warn "Listening on '", join(q{:} => scalar gethostbyaddr($addr, Socket::AF_INET), $port), "'\n"
    if $self->debug;
  $self->_set_port( $relayd->port );
  return;
};

event 'relayd_connected' => sub {
  my ($kernel,$self,$id,$ip) = @_[KERNEL,OBJECT,ARG0,ARG1];
  return;
};

event 'relayd_disconnected' => sub {
  my ($kernel,$self,$id,$ip) = @_[KERNEL,OBJECT,ARG0,ARG1];
  my $data = delete $self->_requests->{$id};
  my $report = eval { Storable::thaw($data); };
  if ( defined $report and ref $report and ref $report eq 'HASH' ) {
    $kernel->yield( 'process_report', $report, $ip );
  }
  else {
    return unless $self->debug;
    warn "Client '$id' failed to send parsable data!\n";
    warn "The error from Storable::thaw was '$@'\n";
  }
  return;
};

event 'relayd_client_input' => sub {
  my ($kernel,$self,$id,$data) = @_[KERNEL,OBJECT,ARG0,ARG1];
  $self->_requests->{$id} .= $data;
  return;
};

event 'process_report' => sub {
  my ($kernel,$self,$data,$ip) = @_[KERNEL,OBJECT,ARG0,ARG1];
  my @present = grep { defined $data->{$_} } @fields;
  return unless scalar @present == scalar @fields;
  # Build CPAN::Testers::Report with its various component facts.
  my $metabase_report = eval { CPAN::Testers::Report->open(
    resource => 'cpan:///distfile/' . $data->{distfile}
  ); };

  return unless $metabase_report;

  $kernel->post( $self->session, $self->recv_event, $data, $ip )
    if $self->recv_event;

  $metabase_report->add( 'CPAN::Testers::Fact::LegacyReport' => {
    map { ( $_ => $data->{$_} ) } qw(grade osname osversion archname perl_version textreport)
  });

  # TestSummary happens to be the same as content metadata
  # of LegacyReport for now
  $metabase_report->add( 'CPAN::Testers::Fact::TestSummary' =>
    [$metabase_report->facts]->[0]->content_metadata()
  );

  $metabase_report->close();

  $kernel->yield( 'submit_report', $metabase_report );
  return;
};

event 'submit_report' => sub {
  my ($kernel,$self,$report) = @_[KERNEL,OBJECT,ARG0];
  $kernel->post(
    $self->queue->get_session_id,
    'submit',
    $report,
  );
  return;
};

sub _load_id_file {
  my $self = shift;

  open my $fh, '<', $self->id_file
    or Carp::confess __PACKAGE__. ": could not read ID file '" . $self->id_file
    . "'\n$!";

  my $data = JSON->new->decode( do { local $/; <$fh> } );

  my $profile = eval { Metabase::User::Profile->from_struct($data->[0]) }
    or Carp::confess __PACKAGE__ . ": could not load Metabase profile\n"
    . "from '" . $self->id_file . "':\n$@";

  my $secret = eval { Metabase::User::Secret->from_struct($data->[1]) }
    or Carp::confess __PACKAGE__ . ": could not load Metabase secret\n"
    . "from '" . $self->id_file . "':\n $@";

  $self->_set_profile( $profile );
  $self->_set_secret( $secret );
  return 1;
}

no MooseX::POE;

__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

POE::Component::Metabase::Relay::Server - A Metabase relay server component

=head1 VERSION

version 0.34

=head1 SYNOPSIS

  use strict;
  use warnings;

  use POE qw[Component::Metabase::Relay::Server];

  my $test_httpd = POE::Component::Metabase::Relay::Server->spawn(
    port    => 8080,
    id_file => shift,
    dsn     => 'dbi:SQLite:dbname=dbfile',
    uri     => 'https://metabase.example.foo/',
    debug   => 1,
  );

  $poe_kernel->run();
  exit 0;

=head1 DESCRIPTION

POE::Component::Metabase::Relay::Server is a relay server for L<Metabase>. It provides a listener
that accepts connections from L<Test::Reporter::Transport::Socket> based CPAN Testers and
relays the L<Storable> serialised data to L<Metabase> using L<POE::Component::Metabase::Client::Submit>.

L<POE::Component::Client::HTTP> is used to submit reports usually, but if version C<0.06> of
L<POE::Component::Curl::Multi> is found to be installed, this will be used in preference. You can
disable this usage using the C<no_curl> option to C<spawn>.

=for Pod::Coverage START

=head1 CONSTRUCTOR

=over

=item C<spawn>

Spawns a new component session and creates a SQLite database if it doesn't already exist.

Takes a number of mandatory parameters:

  'id_file', the file path of a Metabase ID file;
  'dsn', a DBI DSN to use to store the submission queue;
  'uri', the uri of metabase server to submit to;

and a number of optional parameters:

  'address', the address to bind the listener to, defaults to INADDR_ANY;
  'port', the port to listen on, defaults to 0, which picks a random port;
  'username', a DSN username if required;
  'password', a DSN password if required;
  'db_opts', a hashref of DBD options that is passed to POE::Component::EasyDBI;
  'debug', enable debugging information;
  'multiple', set to true to enable the Queue to use multiple PoCo-Client-HTTPs, default 0;
  'no_relay', set to true to disable report submissions to the Metabase, default 0;
  'no_curl',  set to true to disable automatic usage of POE::Component::Curl::Multi, default 0;
  'submissions', an int to control the number of parallel http clients ( used only if multiple == 1 ), default 10;
  'session', a POE::Session alias or session ID to send events to;
  'recv_event', an event to be triggered when reports are received by the relay;

C<address> may be either an simple scalar value or an arrayref of addresses to bind to.

If C<recv_event> is specified an event will be sent for every report received by the relay server.
Unless C<session> is specified this event will be sent to the parent session of the component.

=back

=head1 OUTPUT EVENTS

If C<recv_event> is specified to C<spawn>, an event will be sent with the following:

C<ARG0> will be a C<HASHREF> with the following keys:

 osversion
 distfile
 archname
 textreport
 osname
 perl_version
 grade

C<ARG1> will be the IP address of the client that sent the report.

If C<queue_event> is specified to C<spawn>, an event will be sent for particular changes in queue status

=head1 AUTHOR

Chris Williams <chris@bingosnet.co.uk>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Chris Williams.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
